### ISSUE TITLE
Final and child classes

### ISSUE THREAD

**Issue by kLabz** [HaxeFoundation/haxe#9552]
With 4.1.x, I cannot use `final` on a field that is meant to be set in child classes' constructor (`ReactComponent`'s `state` in react):

```haxe
class Main extends Base {
	public function new() {
		super();
		test = true; // Cannot access field or identifier test for writing
	}

	static function main() {
		new Main();
	}
}

class Base {
	final test:Bool;
	function new() {
		test = false; // Added so that compiler doesn't complain, but this part is actually not needed/wanted
	}
}
```

What I'd want would be a `final` that only allows assignment in constructors, even if it's a child class' constructor. Note that if it's needed that base class don't assign it for this to work, it's ok for my use case (similar to abstract classes).

Workaround would be a `@:bypassFinal` or something, and I'd add it via macro.

**Comment by nadako** [comment]
> Workaround would be a `@:bypassFinal` or something, and I'd add it via macro.

I think we need that for the reasons stated in https://github.com/HaxeFoundation/haxe/issues/8263#issuecomment-501044107.
TLDR: I believe it should be possible to do:
```haxe
function deserializeMyClass() @:bypassFinal {
  var instance = Type.createEmptyInstance(MyClass);
  instance.finalField = deserializeString();
  return instance;
}
```

**Comment by nadako** [comment]
Some more thoughts on this: apart from explicit `@:bypassFinal` I think that double assignment to `final` fields should be always forbidden and that includes cases with multiple assignments in the same constructor like #9313. I think `final` can be only assigned once per code path.
